src/structures/ApplicationCommand.js

```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+const Base = require('./Base');
+const { ApplicationCommandOptionType } = require('../util/Constants');
+const Snowflake = require('../util/Snowflake');
+
+/**
+ * Represents an application command, see {@link InteractionClient}.
+ * @extends {Base}
+ */
+class ApplicationCommand extends Base {
+  constructor(client, data, guildID) {
+    super(client);
+
+    /**
+     * The ID of the guild this command is part of, if any.
+     * @type {?Snowflake}
+     * @readonly
+     */
+    this.guildID = guildID ?? null;
+
+    this._patch(data);
+  }
+
+  _patch(data) {
+    /**
+     * The ID of this command.
+     * @type {Snowflake}
+     * @readonly
+     */
+    this.id = data.id;
+
+    /**
+     * The ID of the application which owns this command.
+     * @type {Snowflake}
+     * @readonly
+     */
+    this.applicationID = data.application_id;
+
+    /**
+     * The name of this command.
+     * @type {string}
+     * @readonly
+     */
+    this.name = data.name;
+
+    /**
+     * The description of this command.
+     * @type {string}
+     * @readonly
+     */
+    this.description = data.description;
+
+    /**
+     * The options of this command.
+     * @type {Object[]}
+     * @readonly
+     */
+    this.options = data.options?.map(function m(o) {
+      return {
+        type: ApplicationCommandOptionType[o.type],
+        name: o.name,
+        description: o.description,
+        default: o.default,
+        required: o.required,
+        choices: o.choices,
+        options: o.options?.map(m),
+      };
+    });
+  }
+
+  /**
+   * The timestamp the command was created at.
+   * @type {number}
+   * @readonly
+   */
+  get createdTimestamp() {
+    return Snowflake.deconstruct(this.id).timestamp;
+  }
+
+  /**
+   * The time the command was created at.
+   * @type {Date}
+   * @readonly
+   */
+  get createdAt() {
+    return new Date(this.createdTimestamp);
+  }
+
+  /**
+   * Edit this command.
+   * @param {ApplicationCommandOptions} data The data to update the command with
```

Comment:
I feel like this should be a separate type since all of the params are optional and throws away data required by the type

Reply:
alternatively use `Partial<ApplicationCommandOptions>` as was going to be documented by https://github.com/discordjs/discord.js/pull/5144 but document it here instead

Reply:
devsnek/discord.js#13 creates a new typedef for that

Reply:
the typedef created there still does not respect edit data being optional, while options becomes optional, name and description still are not.